fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1529,58 +1560,36 @@ def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
     def local_metadata_dict(self) -> Dict[str, Any]:
         """
         Get the information needed to reconstruct the model from shards offline.
-        """
-
-        params_metadata = []
 
+        See the `consolidate_shard_weights` method below.
+        """
+        param_metadata = []
         for path, m in self.named_modules():
-            if not isinstance(m, FullyShardedDataParallel):
-                continue
-
-            # Dealing with FSDP(flatten_parameter=False)
-            # There are as many sharded parameters as there parameters in the
-            # consolidated model, so we only need to export how to reshape the
-            # parameters to their orginal shape and take care of the padding
-            if not m.flatten_parameters:
-                params_metadata.append(
-                    {
-                        "fsdp_path": _clean_path(path),
-                        "is_flat": False,
-                        "num_padded": m.numel_padded_per_param,
-                        "param_names": [_clean_path(p) for p in m.param_paths],
-                        "param_shapes": [p._orig_size for p in m.params],
-                        "param_numels": [_numel_from_size(p._orig_size) for p in m.params],
-                        "no_broadcast_optim_state": m.no_broadcast_optim_state,
-                    }
-                )
-
-            # Dealing with FSDP(flatten_parameter=True)
-            # Now, there is just one flattened parameter mapped to N different
-            # parameters, so we need to export additional information (numels)
-            # on how to split the "merged" parameters, by extracting the meta-data
-            # used in the FlattenParamsWrapper
-            else:
-                param_names = []
-                for module_path, param_name in m.param_path_infos:
-                    full_param_path = module_path + "." + param_name if module_path else param_name
-                    param_names.append(_clean_path(full_param_path))
-                params_metadata.append(
-                    {
-                        "fsdp_path": _clean_path(path),
-                        "is_flat": True,
```

Comment:
Is it an issue that we don't retain the 
`no_broadcast_optim_state` and `is_flat` keys?

I am about to integrate `local_metadata_dict`/`consolidate_shard_weights` into `fairseq:gshard` which will get merged into `fairseq:master` soon, so it would be good if there was test coverage for `MixtureOfExperts` in `test_fsdp_metadata.py`.

Reply:
The is_flat entry is no longer needed. The no_broadcast flag was not used by the consolidation function.

Can you share a test case for MoE? I can add it to the unit test and make sure it is covered. For this PR, I made sure the metadata and consolidation functions are consistently updated and unit test was passing. But it is possible moe case is somewhat special?

Reply:
Its broken before your change :( 
https://github.com/facebookresearch/fairscale/pull/747
we definitely need `no_broadcast_optim_state`. We can probably do without `is_flat`.

Reply:
It's also broken for vanilla transformers.

I'm looking at it today, but won't be offended if this gets merged while I'm debugging :)

Reply:
Thanks for the PR with supplied tests! I am not in a hurry to merge. I will try to test my new changes with your tests. If making the current master working with your test is a lot of work, there is no need for you to spend too much time on it since I am changing this code anyway. It is better for me to make the new code work with the supplied test cases.

Reply:
> 
> 
> It's also broken for vanilla transformers because it doesn't consider `shared_param_info`.
> 
> I'm looking at it today, but won't be offended if this gets merged while I'm debugging :)

That's a great point on shared_param_info. It is now part of the FlatParameter class and FPW proxy the information to the caller. It should be similar to how I handle the param_info data structure. You can directly work on my branch here if you prefer that. I do think my current code is cleaner than the master version. :-)

Reply:
Lol I read this after I fixed :(

Reply:
pushed fixes here!

Reply:
That's awesome, Sam! Really appreciate the team work here!